node-test/src/test/scala/org/bitcoins/node/networking/peer/DataMessageHandlerTest.scala

```diff
@@ -0,0 +1,177 @@
+    param: FixtureParam =>
+      val SpvNodeConnectedWithBitcoind(spv, bitcoind) = param
+
+      val resultP: Promise[Boolean] = Promise()
+
+      for {
+        // fund bitcoind
+        _ <- bitcoind.getNewAddress.flatMap(
+          bitcoind.generateToAddress(blocks = 101, _))
+
+        sender <- spv.peerMsgSenderF
+
+        txId <- bitcoind.sendToAddress(junkAddress, 1.bitcoin)
+        tx <- bitcoind.getRawTransactionRaw(txId)
+        _ <- bitcoind.generateToAddress(blocks = 1, junkAddress)
+        merkleBlock <- bitcoind.getTxOutProof(Vector(txId))
+
+        payload1 = MerkleBlockMessage(merkleBlock)
+        payload2 = TransactionMessage(tx)
+
+        callback: OnMerkleBlockReceived = (
+            _: MerkleBlock,
+            _: Vector[Transaction]) => {
+          Future {
+            resultP.success(true)
+            ()
+          }
+        }
+
+        callbacks = NodeCallbacks.onMerkleBlockReceived(callback)
+
+        dataMessageHandler = DataMessageHandler(dummyChainApi, callbacks)
+        _ <- dataMessageHandler.handleDataPayload(payload1, sender)
+        _ <- dataMessageHandler.handleDataPayload(payload2, sender)
+        result <- resultP.future
+      } yield assert(result)
+  }
+
+  it must "verify OnBlockReceived callbacks are executed" in {
+    param: FixtureParam =>
+      val SpvNodeConnectedWithBitcoind(spv, bitcoind) = param
+
+      val resultP: Promise[Boolean] = Promise()
+
+      for {
+        // fund bitcoind
+        _ <- bitcoind.getNewAddress.flatMap(
+          bitcoind.generateToAddress(blocks = 101, _))
+
+        sender <- spv.peerMsgSenderF
+
+        hash <- bitcoind.generateToAddress(blocks = 1, junkAddress).map(_.head)
+        block <- bitcoind.getBlockRaw(hash)
+
+        payload = BlockMessage(block)
+
+        callback: OnBlockReceived = (_: Block) => {
+          Future {
+            resultP.success(true)
+            ()
+          }
+        }
+
+        callbacks = NodeCallbacks.onBlockReceived(callback)
+
+        dataMessageHandler = DataMessageHandler(dummyChainApi, callbacks)
+        _ <- dataMessageHandler.handleDataPayload(payload, sender)
+        result <- resultP.future
+      } yield assert(result)
+  }
+
+  it must "verify OnBlockHeadersReceived callbacks are executed" in {
+    param: FixtureParam =>
+      val SpvNodeConnectedWithBitcoind(spv, bitcoind) = param
+
+      val resultP: Promise[Boolean] = Promise()
+
+      for {
+        // fund bitcoind
+        _ <- bitcoind.getNewAddress.flatMap(
+          bitcoind.generateToAddress(blocks = 101, _))
+
+        sender <- spv.peerMsgSenderF
+
+        hash <- bitcoind.generateToAddress(blocks = 1, junkAddress).map(_.head)
+        header <- bitcoind.getBlockHeaderRaw(hash)
+
+        payload = HeadersMessage(CompactSizeUInt.one, Vector(header))
+
+        callback: OnBlockHeadersReceived = (_: Vector[BlockHeader]) => {
+          Future {
+            resultP.success(true)
```

Comment:
in our assertions can you make `resultP` the type of the parameter of the callback. This way we can complete it with the parameter for the callback and we can guarantee that the same thing we expected to be passed in is the thing that the callback is executing on? It's just a bit of a sanity check